### DISCUSSION TITLE
Creating Async Middleware

### DISCUSSION THREAD

**Discussion by maheshrajrp** [pmndrs/zustand#1472] [Q&A]
Hi,
I am planning to create an async middleware.
Consider this example,
I wish to keep `loading` as true, when fetch call is happening. Once it is done it will be reverted back to false. I'll be using this `loading` property in a loader component. Since I currently have around 20 similar fetch calls which will drastically increase, I am planning to take this approach. Kindly suggest if there are other better approaches too. 

```
const loader: LoggerImpl = (f) => (set, get, store) => {
    const loggedSet: typeof set =  (...a) => {
        set({ loading: true })
        console.log("2")
        set(...a)  -------------------------------------> (1)
        set({ loading: false })
        console.log("3")
    }
    store.setState = loggedSet
    return f(loggedSet, get, store)
}

export const useDummyStore = create<DummyStoreType>(loader(((set, get, api) => ({
    loading: false,
    value: 1
    fetchCall: async (val: any) => {
        console.log("1")     
        set(async (prev: any) => {
            await setTimeout(() => {
                console.log("4")
            }, 2000);
            return { value: Math.random() % 10 }
        })
        console.log("5") 
       return "Completed";
    }
}))))
```

I want the logs to in the order 1,2,3,4,5. Currently it get executed in 1,2,3,5,4 Order as middleware doesn't wait for statement marked as (1) to be completed.

Thanks in advance.

**Comment by dai-shi** [comment]
> Kindly suggest if there are other better approaches too.

React will provide first-class promise support in the future. It would help dealing with async.
See the example in this tweet: https://twitter.com/dai_shi/status/1580931674656817153

Not a current solution though.

**Reply by Elvis116** [reply]
Can this be done in v5? I need to create a global loading middleware.

**Reply by dai-shi** [reply]
> Can this be done in v5?

v5 isn't different from v4 in this sense.

> I need to create a global loading middleware.

Interesting idea. How does it look like?

**Comment by Elvis116** [comment]
The loading middleware is to set isLoading status for every action which has http request. We may have several data requests in one page, we can group all of them to a namespace, if any of the request is pending we show the Loader. Hide it when all of them are completed. Do you have any idea to accomplish it in zustand ?﻿

Can this be done in v5?

v5 isn't different from v4 in this sense.

I need to create a global loading middleware.

Interesting idea. How does it look like?

**Reply by dai-shi** [reply]
I think we want to first find a solution without middleware.
Having a store-level loading counter might seem to work.

**Reply by Elvis116** [reply]
Is there a global place to add this function so that we don't need to add loading status everywhere. We just need to set isLoading to true before http request and set to false when the request is resolved or in exception.

**Reply by dai-shi** [reply]
I'm still not sure about your requirement, but something simple would be like this:

```js
const useFooStore = create((set) => ({
  isLoading: 0, // global loading counter
  fetchFoo: async () => {
    set((state) => ({ isLoading: state.isLoading + 1 }));
    // do something
    set((state) => ({ isLoading: state.isLoading - 1 }));
  },
  fetchBar: async () => {
    set((state) => ({ isLoading: state.isLoading + 1 }));
    // do something
    set((state) => ({ isLoading: state.isLoading - 1 }));
  },
}));
```
